packages/livekit-rtc/src/audio_source.ts

```diff
@@ -47,7 +63,49 @@ export class AudioSource {
     this.ffiHandle = new FfiHandle(res.source.handle.id);
   }
 
+  get queuedDuration(): number {
+    return Math.max(
+      this.currentQueueSize - Number(process.hrtime.bigint() / 1000000n) + this.lastCapture,
+      0,
+    );
+  }
+
+  clearQueue() {
+    const req = new ClearAudioBufferRequest({
+      sourceHandle: this.ffiHandle.handle,
+    });
+
+    FfiClient.instance.request<ClearAudioBufferResponse>({
+      message: {
+        case: 'clearAudioBuffer',
+        value: req,
+      },
+    });
+
+    this.releaseQueue.put();
+  }
+
+  async waitForPlayout() {
+    await this.releaseQueue.get().then(() => {
+      this.lastCapture = 0;
+      this.currentQueueSize = 0;
+    });
+  }
+
   async captureFrame(frame: AudioFrame) {
+    const now = Number(process.hrtime.bigint() / 1000000n);
+    const elapsed = this.lastCapture === 0 ? 0 : now - this.lastCapture;
+    this.currentQueueSize += (frame.samplesPerChannel / frame.sampleRate - elapsed) * 1000;
+
+    // remove 50ms to account for processing time (e.g. using wait_for_playout for very small chunks)
+    this.currentQueueSize -= 50;
+    this.lastCapture = now;
+
+    if (this.timeout) {
```

Comment:
Seems like this.timeout is never set?

Is `waitForPlayout` working correctly?

Reply:
oops yeah. it seemed to work fine enough so i didn't notice.